src/pages/workspaces/workspace/analysis/modals/AnalysisDuplicator.test.ts

```diff
@@ -0,0 +1,351 @@
+} from 'src/pages/workspaces/workspace/analysis/file-utils'
+import {
+  AnalysisDuplicator,
+  AnalysisDuplicatorProps
+} from 'src/pages/workspaces/workspace/analysis/modals/AnalysisDuplicator'
+import { cloudProviders } from 'src/pages/workspaces/workspace/analysis/runtime-utils'
+import {
+  getToolFromFileExtension,
+  ToolLabel,
+  toolLabelTypes,
+  tools
+} from 'src/pages/workspaces/workspace/analysis/tool-utils'
+import { cloudProviderTypes } from 'src/pages/workspaces/workspace/workspace-utils'
+
+
+jest.mock('src/components/Modal', () => {
+  const { div, h } = jest.requireActual('react-hyperscript-helpers')
+  const originalModule = jest.requireActual('src/components/Modal')
+  return {
+    ...originalModule,
+    __esModule: true,
+    default: props => div({ id: 'modal-root' }, [
+      h(originalModule.default, { onAfterOpen: jest.fn(), ...props })
+    ])
+  }
+})
+
+jest.mock('src/libs/notifications', () => ({
+  notify: (...args) => {
+    console.debug('######################### notify')/* eslint-disable-line */
+    console.debug({ method: 'notify', args: [...args] })/* eslint-disable-line */
+  }
+}))
+
+jest.mock('src/libs/ajax/GoogleStorage')
+jest.mock('src/libs/ajax/AzureStorage')
+
+const getTestFile = (abs: AbsolutePath): AnalysisFile => ({
+  name: abs,
+  ext: getExtension(abs),
+  displayName: getDisplayName(abs),
+  fileName: getFileName(abs),
+  tool: getToolFromFileExtension(getExtension(abs)) as ToolLabel,
+  lastModified: new Date().getTime(),
+  cloudProvider: cloudProviderTypes[cloudProviders.gcp.label]
+})
+
+const baseTestFile: AnalysisFile = getTestFile('test/file0.ipynb' as AbsolutePath)
+
+const onDismiss = jest.fn()
+const onSuccess = jest.fn()
+
+const defaultModalProps: AnalysisDuplicatorProps = {
+  destroyOld: false,
+  fromLauncher: false,
+  printName: baseTestFile.fileName,
+  toolLabel: toolLabelTypes[tools.Jupyter.label],
+  workspace: defaultGoogleWorkspace.workspace,
+  onDismiss, onSuccess
+}
+
+// eslint-disable-next-line
+export const asMockedFn = <T extends (...args: any[]) => any>(
+  fn: T
+): jest.MockedFunction<T> => {
+  return fn as jest.MockedFunction<T>
+}
+
+describe('AnalysisDuplicator', () => {
+  beforeEach(() => {
+    // asMockedFn(withRequesterPays).mockResolvedValue([])
+
+    // Arrange
+    const googleStorageMock: Partial<GoogleStorageContract> = ({
+      listAnalyses: jest.fn(() => Promise.resolve([]))
+    })
+    asMockedFn(GoogleStorage).mockImplementation(() => googleStorageMock as GoogleStorageContract)
+
+    const azureStorageMock: Partial<AzureStorageContract> = ({
+      listNotebooks: jest.fn()
+    })
+    asMockedFn(AzureStorage).mockImplementation(() => azureStorageMock as AzureStorageContract)
+  })
+
+  it('renders correctly by default with destroyOld false', async () => {
+    await act(async () => {
+      await render(h(AnalysisDuplicator, defaultModalProps))
+    })
+
+    screen.getByText(`Copy "${baseTestFile.fileName}"`)
```

Comment:
Also, I think that @msilva-broad likes this pattern of calling `getByText` to verify an element exists because if it fails you get the current DOM elements returned. That's nice for debugging, but I find that that the tests themselves then don't ready very well… it is not obvious what is actually being tested. I would like to see at least a helper method like `assertTextPresent` (that can then call whatever method necessary to get nice debugging output) so that the test is more readable.

Reply:
actually, screen is recommended as the general pattern:

https://kentcdodds.com/blog/common-mistakes-with-react-testing-library#not-using-screen

Reply:
TIL.

I like @cahrens's suggestion for a helper.